test/framework/src/main/java/org/elasticsearch/test/rest/ESRestTestNodesFeatures.java

```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.test.rest;
+
+public interface ESRestTestNodesFeatures {
```

Comment:
While I understand the desire for type safety, this won't be strongly typed in yaml tests. Plus strongly typing like this means we have to adjust this test only class for every feature, which I think will be cumbersome. I think we can still have safety, just at test runtime, by failing if an unknown feature is requested by a test.

Reply:
> Plus strongly typing like this means we have to adjust this test only class for every feature, which I think will be cumbersome

I think there might be a misunderstanding here. This class is for the test framework (and derived test classes) consumption, so that tests can take different actions/decisions based on the cluster version under test (current or an older version in BwC/upgrade/mixed tests).
So this class contains all and only the features needed by the framework (e.g.: can I run the test with soft delete disabled? Can I reset ML feature states on test cleanup? etc.). The 13 features I have extracted so far are everything is needed by ESRestTests and derived.
Of course, YAML tests are not covered and may need more, and you have a very valid point about these needing a different access (preferably, keyed by string). That makes me lean towards the string based implementation. I just wanted to make it clear what's the scope and purpose of this.
  
> Additionally we have to define the mapping twice. I think we should be re-using the mapping defined for production.

If I get this correctly, you mean that we should use features defined in the codebase, e.g. `static final NodeFeature ESQL_STATS_FEATURE = new NodeFeature("esql.stats_node")`. Since these features will be defined each in its own plugin, that means that we will have to take a dependency on all the plugins containing features we will be interested on and call the various FeatureSpecification methods.
I'm not a big fan of the duplication, but I'm not sure I want to add all these dependencies to the test framework either - again, we need to define just some features needed by the famework here. There will be overlap, but it will not be an exact duplicate.

On the other hand, you convinced me mentioning YAML tests that it's better to have features defined as strings; it would be wise to use the same strings as the production definition. 
I think it's better to define (part of) the mapping twice rather than adding dependencies, but I'm open.

Reply:
> I'm not sure I want to add all these dependencies to the test framework either

These would be loaded through SPI, as they will be for production. While at least for unit tests the plugin code would generally be available (and thus the SPI implementation/registration as well), it is true that running integration tests would typically not have all plugins available within the test runner. I wonder if we could build this up, essentially like part of the yaml spec for the distribution? That is, server+each module/plugin has a build task which extracts the features registered. Then we combine those to say "these are all the features of the distribution" which the test runner uses to validate/use.

> I think it's better to define (part of) the mapping twice rather than adding dependencies

If the only option was to define twice or add dependencies, I would agree, but I think there are other options, and defining the mapping twice is quite cumbersome and error prone.

Reply:
> That is, server+each module/plugin has a build task which extracts the features registered.

That's a very interesting thought. I was playing with the same (or similar) idea in my mind. If we can extract the registered features as part of the build task, those could be dumped somewhere and read by the rest tests. We have to consider bwc tests and the ability to store multiple versions though (I'm not sure where we stand with feature removal, but I remember we want to remove all "previous major" features because we consider them widely available and we want to keep the feature set under control); I discarded the idea because it seemed too complex to me, but maybe I was wrong. It will replace any need to have an additional rest API to expose features just for tests (assuming we do not want to expose features).

Reply:
Back on this PR: I will update it to use string-base tests instead of multiple boolean functions, with a concrete implementation that uses Version from NodeInfo (for now) but can be replaced with any mechanism we decide later (ad-hoc rest API, its own definition(s), output of a build task) by creating a different concrete class. Sounds like a good plan?